azure-kusto-data/azure/kusto/data/_token_providers.py

```diff
@@ -179,19 +209,29 @@ def context(self) -> dict:
 
     def _init_impl(self):
         try:
-            self.msi_auth_context = ManagedIdentityCredential(**self._msi_args)
+            if self.is_async:

```

Comment:
This is odd.
Suggest you lazy init the correct instance upon use instead of keeping this member

Reply:
@yogilad You're right, get_token should always have been protected with I lock.

The problem with lazy-initing get_token_cache for msi is that I'll need to lock it too.